src/redis-cli.c

```diff
@@ -736,8 +737,13 @@ static int cliSwitchProto(void) {
 
     int result = REDIS_OK;
     if (reply->type == REDIS_REPLY_ERROR) {
-        result = REDIS_ERR;
-        fprintf(stderr,"HELLO 3 failed: %s\n",reply->str);
+        if (config.output == OUTPUT_JSON) {
```

Comment:
it's a bit odd to have the check for OUTPUT_JSON check here.
maybe we can document near the declaration of `config.resp3`, that a value of 2 mean that it's implicit.
then here, we print just one error, but if config.resp3 is 1, we return an error, and if it's 2, we return ok (not explicitly checking for JSON)

Reply:
I realized that a user can get a result as json like the implicit mode with `-3` and `--json` option in REPL mode and RESP2 server.

Redis server version is 3.2.13
```
$ src/redis-cli -3
HELLO 3 failed: ERR unknown command 'HELLO'
127.0.0.1:6379> hgetall user1
1) "id"
2) "1"
3) "age"
4) "22"

$ src/redis-cli -3 --json
HELLO 3 failed: ERR unknown command 'HELLO'
127.0.0.1:6379> hgetall user1
["id","1","age","22"]
```

Reply:
i'm not sure i understand your comment.
if the user added an explicit `-3` then he deserves the error message, and if all he did was add `--json`, then he won't get it.
that's the same in interactive and non-interactive modes.
am i missing anything?

Reply:
I just wonder what the spec is if a user add an explicit `-3` in REPL which version only supports RESP2.

In non-interactive mode, a user will get only the error message, not get a command result regardless of whether or not`--json` is
```
$ src/redis-cli -3 hgetall user1
HELLO 3 failed: ERR unknown command 'HELLO'

$ src/redis-cli -3 --json hgetall user1
HELLO 3 failed: ERR unknown command 'HELLO'
```

In interactive mode, a user will get the error message, in subsequent command, a user can get a command result regardless of whether or not `--json` is.

```
$ src/redis-cli -3
HELLO 3 failed: ERR unknown command 'HELLO'
127.0.0.1:6379> hgetall user1
1) "id"
2) "1"
3) "age"
4) "22"

$ src/redis-cli --json
HELLO 3 failed: ERR unknown command 'HELLO'
127.0.0.1:6379> hgetall user1
["id","1","age","22"]
```

@oranagra This is a spec, then can I keep as is?

Reply:
so in non-interactive mode this is an **error** and return code is 1 (error).
and in interactive mode, this is a warning, and the user can chose to ignore it.
i think this is ok, considering that non-interactive mode is also used for script (non human).
and interactive mode is used by humans.

however, if the user just provides `--json` (and the RESP3 part is implicit), then even in non-interactive mode it's only considered as a warning, and exit code is 0 (success).
it does print a warning message, which can be silenced by adding `-2`

so the current code seems fine to me. i'll describe it in the top comment.